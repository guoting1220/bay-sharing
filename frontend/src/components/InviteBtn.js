import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { sendInvite, removeInvite } from '../actions/userActions';

const InviteBtn = ({ postId, postOwner }) => {
  const currentUserName = useSelector(store => store.user.username);

  const sentInvitePostIds = (useSelector(store => store.user.sentInvites)).map(invite => invite.postId);
  const sentIdSet = new Set(sentInvitePostIds);

  const dispatch = useDispatch();

  /* the text and style of the invite button is decided whether the post 
  is invited or not */
  const inviteBtn = sentIdSet.has(postId) ?
    <div className="btn btn-sm btn-success" data-toggle="tooltip" title="cancel invite">Invite Sent</div>
    :
    <div className="btn btn-sm btn-outline-success" data-toggle="tooltip" title="send invite">Invite</div>

  const toggleInvite = () => {
    if (!sentIdSet.has(postId)) {
      dispatch(sendInvite(currentUserName, postId, postOwner));
    }
    else
      dispatch(removeInvite(currentUserName, postId));
  }

  /* the invite button is hidden for the current user's posts 
  (the user can not invite themselve's posts) */
  if (postOwner === currentUserName) return null;

  return (
    <div className="InviteBtn" onClick={toggleInvite}>
      {inviteBtn}
    </div>
  )
}

export default InviteBtn;