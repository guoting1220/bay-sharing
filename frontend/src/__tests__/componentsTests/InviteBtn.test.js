import React from "react";
import InviteBtn from "../../components/InviteBtn";
import { MemoryRouter } from "react-router";
import { wraperRender } from './testUtils';
import { screen } from '@testing-library/react'

describe("render InviteBtn component", () => {
  it("renders without crashing", function () {
    wraperRender(
      <MemoryRouter>
        <InviteBtn postId={1} postOwner={"b"}/>
      </MemoryRouter>,
      {
        initialState: {
          user: {
            username: "a",
            sentInvites: [{ postId: 99, postOwner: "b" }]
          }
        }
      }
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = wraperRender(
      <MemoryRouter>
        <InviteBtn postId={1} postOwner={"b"}/>
      </MemoryRouter>,
      {
        initialState: {
          user: {
            username: "a",
            sentInvites: [{ postId: 99, postOwner: "b" }]
          }
        }
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with invite button", function () {
    wraperRender(
      <MemoryRouter>
        <InviteBtn postId={1} postOwner={"b"} />
      </MemoryRouter>,
      {
        initialState: {
          user: {
            username: "a",
            sentInvites: [{ postId: 99, postOwner: "b" }]
          }
        }
      }
    );
    expect(screen.queryByText("Invite")).toBeInTheDocument();
  });

  it("renders with invite sent button", function () {
    wraperRender(
      <MemoryRouter>
        <InviteBtn postId={99} postOwner={"b"} />
      </MemoryRouter>,
      {
        initialState: {
          user: {
            username: "a",
            sentInvites: [{ postId: 99, postOwner: "b" }]
          }
        }
      }
    );
    expect(screen.queryByText("Invite Sent")).toBeInTheDocument()
  });

})