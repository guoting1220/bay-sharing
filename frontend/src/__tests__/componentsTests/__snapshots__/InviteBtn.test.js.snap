// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render InviteBtn component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="InviteBtn"
  >
    <div
      class="btn btn-sm btn-outline-success"
    >
      Invite
    </div>
  </div>
</DocumentFragment>
`;
